### ISSUE TITLE
Replace the class BrowseProviderFileCaching with Rails.cache

### ISSUE THREAD

**Issue by prioux** [aces/cbrain#1094]
The class BrowseProviderFileCaching just fetches and stores browsing data in a file in /tmp ; replace the caching part with calls to Rails.cache instead.

Make sure the cache key is tied to all three of : 1) which user is browsing, 2) which data provider is being browsed, and 3) under which path is the browsing happening.

**Comment by MontrealSergiy** [comment]
@prioux  Is production backend cache configured to keep large values? Anything needs to be addressed? 

https://stackoverflow.com/questions/37352553/how-to-increase-memcache-slab-size-above-1mb-with-dalli-and-rails

**Comment by prioux** [comment]
I have not set any particular value for that in the Rails code. Wher eis this configured? Do you think it might be an issue?

**Comment by prioux** [comment]
Assume the browsing module will cache up to 10,000 entries; estimate the size of one entry (or generate a cache with 100 entries and measure it by looking ta the file) and let me know how we should adjust the limit.

**Comment by prioux** [comment]
Wait, let's increase it to 50,000 entries. So we can browse the full set of UK Biobank subjects.

**Comment by MontrealSergiy** [comment]
Should say I do not know for sure what is used on production to cache (but MemCache is popular choice) and is there any problem, Rails, for example circumvents the MemCache limit of key size. The stackoverflow question referred seems to be about a specific gem.

I will see now what's size cache is on my pc per file but it my vary  depending on length of file name. Upper limit but 50000 * 256 (max length of a path in linux) = 12500000  = 12 M, while for MemCache the max value is 1Mb by default. Paths are even longer in linux (up to 4M)

**Comment by prioux** [comment]
We're not using the mem cache store in production, we're using the disk cache store, otherwise the caches wouldn't be propagated from one instance to another. Just cache something with the console and find the file on disk, like I said.

**Comment by prioux** [comment]
To test:

* Start a Rails console started in production mode (because the Rails cache configuration is not the same in developement):

```
rails console production)
```

* Create a dummy FileInfo object and an array of 100 of them:

```ruby
fi = FileInfo.new(:name => "x" * 100, :symbolic_type => :regular, :atime => Time.now)
ar = 100.times.map { fi.dup };1
```

* Trigger caching of the array:

```ruby
res = Rails.cache.fetch("mykey") { ar };1
```

* Find the cache file in a shell:

```
cd BrainPortal/tmp/cache
find . -type f -mmin -5 -print # any file modified in the past 5 minutes
```

Examine its size etc.

**Comment by MontrealSergiy** [comment]
Good to know. BTW I do not think MemCache or Redis is per instance (not 100% sure either) 




i

**Comment by prioux** [comment]
I don't know, if the mem cache is shared between processes, it means the processes have to establish a communication channel between themselves. And I'm not talking about separate puma thread within the same process, I'm talking about separate unix processes. I don't know how the processes would find each other.

**Comment by prioux** [comment]
The production file system is plenty fast enough for us anyway.

**Comment by MontrealSergiy** [comment]
Even with short names, 10000 files take over 1Mb. ( 1838486 b). But seems fine for disk based store.  May be some comment in prod config file (if use advance in memory caching mode, configure to hold large values, depending on size of dp's).

**Comment by prioux** [comment]
That's nothing. For 50,000 files that means 5Mb. It's insignificant, and it will all resides in the kernel's disk cache anyway.

**Comment by prioux** [comment]
Fixed by #1099
